src/nbl/video/COpenGL_Swapchain.cpp

```diff
@@ -0,0 +1,414 @@
+#include "nbl/video/COpenGL_Queue.h"
+
+#include "nbl/system/IThreadHandler.h"
+
+namespace nbl::video
+{
+
+static inline constexpr uint32_t OpenGLFunctionTableSize = 6672u;
+using SThreadHandlerInternalState = std::array<uint8_t, OpenGLFunctionTableSize>;
+
+static_assert(OpenGLFunctionTableSize >= sizeof(COpenGLFunctionTable));
+static_assert(OpenGLFunctionTableSize >= sizeof(COpenGLESFunctionTable));
+
+IOpenGL_FunctionTable* getFunctionPointer(video::E_API_TYPE apiType, SThreadHandlerInternalState* internalState)
+{
+    if (apiType == video::EAT_OPENGL)
+        return reinterpret_cast<COpenGLFunctionTable*>(internalState);
+    else if (apiType == video::EAT_OPENGL_ES)
+        return reinterpret_cast<COpenGLESFunctionTable*>(internalState);
+    else assert(false);
+}
+
+class COpenGL_SwapchainThreadHandler final : public system::IThreadHandler<COpenGL_SwapchainThreadHandler, SThreadHandlerInternalState>
+{
+    friend COpenGLSwapchain;
+    friend COpenGLESSwapchain;
+    static inline constexpr uint32_t MaxImages = 4u;
+    using base_t = system::IThreadHandler<COpenGL_SwapchainThreadHandler, SThreadHandlerInternalState>;
+    friend base_t;
+
+public:
+    COpenGL_SwapchainThreadHandler(const egl::CEGL* _egl,
+        core::smart_refctd_ptr<IOpenGL_LogicalDevice> dev,
+        const void* _window,
+        ISurface::E_PRESENT_MODE presentMode,
+        uint32_t _imgCount, uint32_t _imgWidth, uint32_t _imgHeight,
+        const COpenGLFeatureMap* _features,
+        EGLContext _ctx, EGLSurface _surface,
+        EGLConfig _config,
+        COpenGLDebugCallback* _dbgCb
+    ) : m_device(dev), m_masterContextCallsWaited(0),
+        egl(_egl),
+        m_presentMode(presentMode),
+        glctx{ _ctx,_surface },
+        features(_features),
+        imageCount(_imgCount),
+        imgWidth(_imgWidth),
+        imgHeight(_imgHeight),
+        m_dbgCb(_dbgCb)
+    {}
+
+    void requestBlit(uint32_t _imgIx, uint32_t semCount, IGPUSemaphore* const* const sems)
+    {
+        auto raii_handler = base_t::createRAIIDispatchHandler();
+
+        needToBlit = true;
+        request.imgIx = _imgIx;
+        request.semCount = semCount;
+        request.sems = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<core::smart_refctd_ptr<COpenGLSemaphore>>>(semCount);
+        for (uint32_t i = 0u; i < semCount; ++i)
+        {
+            COpenGLSemaphore* sem = IBackendObject::device_compatibility_cast<COpenGLSemaphore*>(sems[i], m_device.get());
+            request.sems->begin()[i] = core::smart_refctd_ptr<COpenGLSemaphore>(sem);
+        }
+    }
+
+    core::smart_refctd_ptr<COpenGLSync> getSyncForImgIx(uint32_t imgix)
+    {
+        auto lk = base_t::createLock();
+
+        return syncs[imgix];
+    }
+
+    void createImgViewFbo(uint32_t imgix, core::smart_refctd_ptr<IGPUImage> image)
+    {
+        auto lk = base_t::createRAIIDispatchHandler();
+        request.initializeImg = image;
+        request.imgIx = imgix;
+    }
```

Comment:
don't create them on demand, create them up-front, along with image `GLuint` for later use with `IGPUImage` on-demand creation (basically same pattern as Vulkan, some images that need to get retrieved)